docs/design/dft.md

```diff
@@ -0,0 +1,1007 @@
+layout of this idea. The Scaled Endpoint operator takes the Scaled
+Endpoint configuration - written in a high level language - and
+generates a specification table. A function verifier, such as Gravel
+, takes the specification table and predefined
+specification of the middleboxes as inputs and provide a proof of the
+correctness of the Scaled Endpoint configuration. If proved correct,
+the Scaled Endpoint propagates the specification table to
+the Scaled Endpoint's FWDs and DFT.
+
+# DFT Internal Protocol
+
+**NSH Metadata:** The following diagram illustrates the NSH
+metadata structure. We designed the metadata primarily around the DFT
+operations to convey the needed changes that FWDs must apply to
+packets' headers. Figure [8] show the structure of the
+NSH metadata, with fields:
+
+<img src="png/nsh.png" width="600" height="150">
+
+Figure 8: DFT NSH metadata
+
+-   Ver (3-bit): Metadata version
+
+-   Length (5-bits): Total length of the metadata in 4 bytes increments
+
+-   (R)everse (1-bits): Indicates DFT operations for the reverse flow
+
+-   Reserved (23-bits): For future use and must be zero
+
+-   FWD IP Address (32-bits): IP address of FWD to return the packet to
+
+-   DFT operations: Variable length table operations
+
+The DFT NSH metadata allows packing multiple DFT
+operations in the same metadata header as long as there is a room in
+the packet permitted by the network MTU. Figure [9]
+shows the header structure of a DFT operation, with fields:
+
+<img src="png/dftop.png" width="600" height="170">
+
+Figure 9: DFT operation
+
+-   Op (4-bit): Metadata version
+
+  -   0x0: Reserved
+
+  -   0x1: Lookup
+
+  -   0x2: Reply
+
+  -   0x3: Update ANY
+
+  -   0x4: Update NOEXIST
+
+  -   0x5: Update EXIST
+
+  -   0x6: Delete
+
+  -   0x7-0xf: Reserved
+
+-   Index (16-bits): Index of the ebpf map of the FTN
+
+-   Key Len (5-bits): Key length in multiple of 4 bytes
+
+-   Val Len (5-bits): Value length in multiple of 4 bytes
+
+-   Key: Key payload of length (Key Len)
+
+-   Value: Value payload of value (Val Len)
+
+The maximum length of the key and value fits an IP + TCP headers. We
+preserve the exact structure of the key and value fields as they are
+related to the key and value structures of the ebpf map referred to by
+Index. Typically, we shall use an optimized key as reference fields to
+the inner packet header to save space in the metadata header and avoid
+copying unnecessary information that already exist in the packet.
+
+<img src="png/packetflow.png" width="500" height="300">
+
+Figure 10: Regular packet flow
+
+**Packet Flow:** Algorithm 1 and Algorithm
+2 detail the FWD and FTN behaviors for packet
+processing. Figure [10] describes a typical example of
+the regular packet flow and interaction between the FWDs and the DFT.
+The packet progresses from source to destination as follows:
+
+1. FWD receives a packet from source
+2. FWD lookup the packet's flow entry in its flow f-map (cached)
```

Comment:
FWD lookup the packet's flow entry in its flow f-map (cached) -> FWD does the packet's flow entry lookup in its flow f-map (cached)